From 77ad5b89b723be471807f12efee8ea10abe95f4f Mon Sep 17 00:00:00 2001
From: Jason Jae Hyun Koh <jason.koh@motorola.com>
Date: Wed, 15 Oct 2014 15:17:15 -0500
Subject: [PATCH 086/959] IKLOCSEN-1738 stml0xx: allow ioctls between 32-bit
 and 64-bit processes

- Added .compat_ioctl in file_operations
- Removed pointer type in IOCTL definitions that caused a size mismatch

Change-Id: I2dd1ef4b87a663e4670c6787c99fe8a7e82387a4
Signed-off-by: Jason Jae Hyun Koh <jason.koh@motorola.com>
Reviewed-on: http://gerrit.mot.com/679431
SLTApproved: Slta Waiver <sltawvr@motorola.com>
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Robert Wiegand <Robert.P.Wiegand@motorola.com>
Reviewed-by: James Rumpler <w12004@motorola.com>
Reviewed-by: Jeffrey Carlyle <jeff.carlyle@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
---
 drivers/misc/stml0xx/stml0xx_flash.c |  3 +++
 include/linux/stml0xx.h              | 19 ++++++++++---------
 2 files changed, 13 insertions(+), 9 deletions(-)

diff --git a/drivers/misc/stml0xx/stml0xx_flash.c b/drivers/misc/stml0xx/stml0xx_flash.c
index a58a40b8968..655eea08396 100644
--- a/drivers/misc/stml0xx/stml0xx_flash.c
+++ b/drivers/misc/stml0xx/stml0xx_flash.c
@@ -585,5 +585,8 @@ const struct file_operations stml0xx_misc_fops = {
 	.owner = THIS_MODULE,
 	.open = stml0xx_misc_open,
 	.unlocked_ioctl = stml0xx_misc_ioctl,
+#ifdef CONFIG_COMPAT
+	.compat_ioctl = stml0xx_misc_ioctl,
+#endif
 	.write = stml0xx_misc_write,
 };
diff --git a/include/linux/stml0xx.h b/include/linux/stml0xx.h
index d24844f664f..d03718e6783 100644
--- a/include/linux/stml0xx.h
+++ b/include/linux/stml0xx.h
@@ -112,13 +112,13 @@
 #define STML0XX_IOCTL_SET_POWER_MODE	\
 		_IOW(STML0XX_IOCTL_BASE, 27, unsigned char)
 #define STML0XX_IOCTL_GET_ALGOS	\
-		_IOR(STML0XX_IOCTL_BASE, 28, char*)
+		_IOR(STML0XX_IOCTL_BASE, 28, char[STML0XX_ALGO_SIZE])
 #define STML0XX_IOCTL_SET_ALGOS	\
-		_IOW(STML0XX_IOCTL_BASE, 29, char*)
+		_IOW(STML0XX_IOCTL_BASE, 29, char[STML0XX_ALGO_SIZE])
 #define STML0XX_IOCTL_GET_MAG_CAL \
-		_IOR(STML0XX_IOCTL_BASE, 30, unsigned char*)
+		_IOR(STML0XX_IOCTL_BASE, 30, char[STML0XX_MAG_CAL_SIZE])
 #define STML0XX_IOCTL_SET_MAG_CAL \
-		_IOW(STML0XX_IOCTL_BASE, 31, unsigned char*)
+		_IOW(STML0XX_IOCTL_BASE, 31, char[STML0XX_MAG_CAL_SIZE])
 /* 32 unused */
 #define STML0XX_IOCTL_SET_MOTION_DUR	\
 		_IOW(STML0XX_IOCTL_BASE, 33, unsigned int)
@@ -130,20 +130,20 @@
 #define STML0XX_IOCTL_SET_WAKESENSORS	\
 		_IOW(STML0XX_IOCTL_BASE, 37, unsigned char)
 #define STML0XX_IOCTL_GET_VERNAME	\
-		_IOW(STML0XX_IOCTL_BASE, 38, char*)
+		_IOW(STML0XX_IOCTL_BASE, 38, char[FW_VERSION_SIZE])
 #define STML0XX_IOCTL_SET_POSIX_TIME	\
 		_IOW(STML0XX_IOCTL_BASE, 39, unsigned long)
 /* 40-42 unused */
 #define STML0XX_IOCTL_SET_ALGO_REQ \
-		_IOR(STML0XX_IOCTL_BASE, 43, char*)
+		_IOR(STML0XX_IOCTL_BASE, 43, char[1])
 #define STML0XX_IOCTL_GET_ALGO_EVT \
-		_IOR(STML0XX_IOCTL_BASE, 44, char*)
+		_IOR(STML0XX_IOCTL_BASE, 44, char[1])
 #define STML0XX_IOCTL_ENABLE_BREATHING \
 		_IOW(STML0XX_IOCTL_BASE, 45, unsigned char)
 #define STML0XX_IOCTL_WRITE_REG \
-		_IOR(STML0XX_IOCTL_BASE, 46, char*)
+		_IOR(STML0XX_IOCTL_BASE, 46, char[1])
 #define STML0XX_IOCTL_READ_REG \
-		_IOR(STML0XX_IOCTL_BASE, 47, char*)
+		_IOR(STML0XX_IOCTL_BASE, 47, char[1])
 /* 48-52 unused */
 #define STML0XX_IOCTL_GET_BOOTED \
 		_IOR(STML0XX_IOCTL_BASE, 53, unsigned char)
@@ -160,6 +160,7 @@
 #define STML0XX_MAG_CAL_SIZE 26
 #define STML0XX_AS_DATA_BUFF_SIZE 20
 #define STML0XX_MS_DATA_BUFF_SIZE 20
+#define STML0XX_ALGO_SIZE	2
 
 #define STML0XX_CAMERA_DATA 0x01
 
-- 
2.11.0

